### PR TITLE
[TPU][Quantization] TPU `W8A8`

### PR DESCRIPTION
**Pull request by robertgshaw2-redhat**
#### SUMMARY:
* support TPU for `compressed-tensors` w8a8 models.
* To run, just load a W8A8 model:

```python
from vllm import LLM
model = LLM("neuralmagic/Meta-Llama-3.1-8B-Instruct-quantized.w8a8", max_model_len=2048)
model.generate("Hello my name is")
```

#### TESTING:
- [x] verified accuracy on TPU for Llama-8B on TP=1 (exact score as GPU)
- [x] verified accuracy on TPU for Llama-8B on TP=4 (exact score as GPU)
- [x]  verified accuracy on TPU for Llama-70B on TP=1 (exact score as GPU)
- [x] verified accuracy on TPU for Qwen on TP=1 (exact score as GPU)  --- note: bias in model
- [x] confirmed all schemes still work on GPU, including:
    * `nm-testing/Meta-Llama-3-8B-Instruct-W8A8-Dynamic-Asym`
    * `nm-testing/Meta-Llama-3-8B-Instruct-W8A8-Static-Per-Tensor-Sym`
    * `nm-testing/Meta-Llama-3-8B-Instruct-W8A8-Static-Per-Tensor-Asym`
- [x] add Llama TP=1 tests to CI/CD
    * FOLLOW UP: Add more than one model once we enable lm-eval framework on TPU
    * FOLLOW UP: add TP>1 once we enable this machine type in the CI
- [x] figure out workaround for user warning re: cond

#### FOLLOW UP
* [TPU] Mixed precision
* [TPU] Estimated memory usage is elevated due to `peak_bytes` capturing some intermediate tensors, fix it.
* [Software Quality] Add `TritonScaledMMLinear` abstraction
* [Software Quality] Convert Fp8 methods to use Kernel abstraction

### CODE REVIEW

File: vllm/model_executor/layers/quantization/kernels/scaled_mm/xla.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+    def apply_weights(self,
+                      layer: torch.nn.Module,
+                      x: torch.Tensor,
+                      bias: Optional[torch.Tensor] = None) -> torch.Tensor:
+        w_q, w_s, i_s, i_zp, i_azp_adj = self._get_weight_params(layer)
+        assert i_s is None
+        assert i_zp is None
+        assert i_azp_adj is None
+        assert bias is None, "Bias is not supported for XLA yet."
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/kernels/scaled_mm/xla.py:74]: Please apply the bias manually after `torch.ops.xla.quantized_matmul`

File: vllm/model_executor/layers/quantization/kernels/scaled_mm/xla.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            torch.nn.Parameter(weight_scale.data, requires_grad=False))
+
+    def apply_weights(self,
+                      layer: torch.nn.Module,
+                      x: torch.Tensor,
+                      bias: Optional[torch.Tensor] = None) -> torch.Tensor:
+        w_q, w_s, i_s, i_zp, i_azp_adj = self._get_weight_params(layer)
+        assert i_s is None
+        assert i_zp is None
+        assert i_azp_adj is None
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/kernels/scaled_mm/xla.py:73]: nit: you could check these in `process_weights_after_loading`

File: vllm/model_executor/layers/quantization/kernels/scaled_mm/__init__.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+from vllm.model_executor.layers.quantization.kernels.scaled_mm.ScaledMMLinearKernel import (  # noqa: E501
+    ScaledMMLinearKernel, ScaledMMLinearLayerConfig)
+from vllm.model_executor.layers.quantization.kernels.scaled_mm.xla import (
+    XLAScaledMMLinearKernel)
+from vllm.platforms import PlatformEnum, current_platform
+
+# in priority/performance order (when available)
+_POSSIBLE_KERNELS: Dict[PlatformEnum, List[Type[ScaledMMLinearKernel]]] = {
+    PlatformEnum.CPU: [CutlassScaledMMLinearKernel],
+    PlatformEnum.CUDA: [CutlassScaledMMLinearKernel],
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/kernels/scaled_mm/__init__.py:15]: I believe `PlatformEnum.ROCM` should be here as well since there is INT8 W8A8 support https://github.com/vllm-project/vllm/pull/9857

File: vllm/model_executor/layers/quantization/kernels/scaled_mm/xla.py (hunk 1 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/kernels/scaled_mm/xla.py:74]: does Qwen have a bias (for testing)

**Comment by bvrockwell** [comment]
cc @lsy323

File: vllm/model_executor/layers/quantization/kernels/scaled_mm/xla.py (hunk 1 above)
Comment (mgoin) [reply vllm/model_executor/layers/quantization/kernels/scaled_mm/xla.py:74]: Yep there are biases on QKV

Screenshot 2025-01-06 at 8 42 32 PM

File: vllm/model_executor/layers/quantization/kernels/scaled_mm/__init__.py (hunk 3 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/kernels/scaled_mm/__init__.py:15]: good catch. this abstraction is much better

File: vllm/model_executor/layers/quantization/kernels/scaled_mm/xla.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+                                             w_q,
+                                             w_s,
+                                             zero_point=None,
+                                             block_size=-1,
+                                             int4_weight=False,
+                                             quantize_activation=True)
+
+        # Explicitly capture control flow to make dynamo happy.
+        # https://pytorch.org/docs/main/generated/exportdb/index.html#cond-branch-class-method # noqa: E501
+        return cond(bias is None, self._no_add_bias, self._add_bias, [out, bias])
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/kernels/scaled_mm/xla.py:93]: @mgoin thoughts on this?

Comment (mgoin) [reply vllm/model_executor/layers/quantization/kernels/scaled_mm/xla.py]: Ah I see why this is needed, sorry this is more difficult than I assumed. This seems like a nice interface if it is stable. Nice find

File: vllm/model_executor/layers/quantization/kernels/scaled_mm/__init__.py (hunk 3 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/kernels/scaled_mm/__init__.py:17]: NOTE: to keep the scope of this PR reasonable, I will add `TritonScaledMM` Kernel as a follow up.

File: vllm/model_executor/layers/quantization/kernels/scaled_mm/xla.py (hunk 4 above)
Comment (robertgshaw2-neuralmagic) [reply vllm/model_executor/layers/quantization/kernels/scaled_mm/xla.py]: Its still a prototype feature

File: vllm/model_executor/layers/quantization/kernels/scaled_mm/xla.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+                                             w_q,
+                                             w_s,
+                                             zero_point=None,
+                                             block_size=-1,
+                                             int4_weight=False,
+                                             quantize_activation=True)
+
+        # Explicitly capture control flow to make dynamo happy.
+        # https://pytorch.org/docs/main/generated/exportdb/index.html#cond-branch-class-method # noqa: E501
+        return cond(bias is None, self.no_add_bias, self.add_bias, [out, bias])
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/kernels/scaled_mm/xla.py:92]: NOTE: putting a binary object here causes torch to specialize on each dimension. This is fine for us, but it logs a lot of warnings. @mgoin do you think we should just suppress?

Comment (mgoin) [reply vllm/model_executor/layers/quantization/kernels/scaled_mm/xla.py:92]: Yeah I think it's fine to pull this out into a variable and suppress - maybe you could filter it through `print_warning_once`

**Comment by robertgshaw2-neuralmagic** [comment]
@mgoin this is ready to go.

File: vllm/model_executor/layers/quantization/kernels/mixed_precision/__init__.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
+from typing import List, Optional, Type
```
Comment (robertgshaw2-neuralmagic) [inline vllm/model_executor/layers/quantization/kernels/mixed_precision/__init__.py:1]: NOTE for reviewer - this file is not changed, it is just moved

**Review by mgoin — APPROVED** [review]
LGTM, excellent work
